File: scripts/import_summary_resumen_2025.py
import argparse
import json
import sqlite3
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_DB_PATH = ROOT / "datos" / "panel.sqlite"
DEFAULT_CUENTAS = [
    "info IMPORTANTE/RESUMEN CUENTAS 2025.json",
    "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN 2025.json",
    "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN.json",
]
DEFAULT_OPERACIONES = [
    "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN 2025.json",
    "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN.json",
]
EMPRESA_ID = "EMPRESA01"
TARGET_YEAR = 2025

KEY_PRINCIPAL_TARGET = "SECCI\u00e0N Principal"
KEY_SECUNDARIA_TARGET = "SECCION Secundaria"
PRINCIPAL_KEYS = [
    "SECCI\u00f3N Principal",
    "SECCI\u00e0N Principal",
    "SECCI\u00f3N",
    "SECCI\u00e0N",
    "SECCION Principal",
    "SECCION",
]
SECUNDARIA_KEYS = [
    "SECCION Secundaria",
    "SECCI\u00f3N Secundaria",
    "SECCI\u00e0N Secundaria",
    "Seccion Secundaria",
    "SECCION",
    "SECCI\u00f3N",
    "SECCI\u00e0N",
]
OPERATION_KEYS = [
    "sum-row",
    "sum-row-sumavarios",
    "sum-row-sumavarios-consolidado",
    "sum-row-operativo",
    "sum-row-operativo-consolidado",
    "result-row",
    "net-row",
    "net-row-adicional",
    "result-net-row",
]


def resolve_path(value: str | None, fallbacks: list[str]) -> Path:
    if value:
        candidate = Path(value).expanduser().resolve()
        if candidate.exists():
            return candidate
        raise FileNotFoundError(f"No existe el archivo {candidate}")
    for fallback in fallbacks:
        candidate = (ROOT / fallback).resolve()
        if candidate.exists():
            return candidate
    raise FileNotFoundError(
        "No se encontr\u00f3 archivo de entrada. Usa --cuentas o --operaciones para definirlo."
    )


def load_json(path: Path):
    with path.open(encoding="utf-8") as fh:
        return json.load(fh)


def clean_text(value):
    if value is None:
        return ""
    return str(value).strip()


def normalize_principal(row):
    for key in PRINCIPAL_KEYS:
        if key in row and row[key]:
            return clean_text(row[key])
    return ""


def normalize_secundaria(row):
    for key in SECUNDARIA_KEYS:
        if key in row and row[key]:
            return clean_text(row[key])
    return ""


def group_accounts(rows):
    grouped = {}
    for row in rows:
        capitulo = clean_text(row.get("CAPITULO"))
        if not capitulo:
            continue
        cuenta = clean_text(row.get("CUENTA"))
        nombre = clean_text(row.get("NOMBRE"))
        if not cuenta:
            continue
        grouped.setdefault(capitulo, []).append(
            {
                "CUENTA": cuenta,
                "NOMBRE": nombre,
                KEY_PRINCIPAL_TARGET: normalize_principal(row),
                KEY_SECUNDARIA_TARGET: normalize_secundaria(row),
            }
        )
    return grouped


def build_operations(raw_ops, hoja_expected):
    hoja_upper = hoja_expected.upper()
    filtrados = [
        op for op in raw_ops if clean_text(op.get("HOJA")).upper() == hoja_upper
    ]
    operaciones = []
    for idx, op in enumerate(filtrados):
        capitulo = clean_text(op.get("CAPITULO"))
        clase = clean_text(op.get("Clase"))
        seccion = clean_text(op.get("SECCION"))
        if not capitulo or not clase or not seccion:
            continue
        signo = -1 if "expense" in clase.lower() else 1
        for tipo_idx, tipo in enumerate(OPERATION_KEYS):
            etiqueta = clean_text(op.get(tipo))
            if not etiqueta:
                continue
            operaciones.append(
                {
                    "capitulo": capitulo,
                    "clase": clase,
                    "seccion": seccion,
                    "tipo": tipo,
                    "label": etiqueta,
                    "signo": signo,
                    "orden": idx * 100 + tipo_idx,
                }
            )
    return operaciones


def insert_cuentas(conn, modulo, grouped, empresa_id, anio):
    conn.execute(
        "DELETE FROM layout_cuentas WHERE empresa_id = ? AND modulo = ? AND anio = ?",
        (empresa_id, modulo, anio),
    )
    conn.execute(
        "DELETE FROM layout_secciones WHERE empresa_id = ? AND modulo = ? AND anio = ?",
        (empresa_id, modulo, anio),
    )
    insert_cuenta_sql = """
        INSERT OR REPLACE INTO layout_cuentas (
            empresa_id, modulo, anio, cuenta, nombre, capitulo,
            seccion_principal, seccion_secundaria, orden, actualizado_en
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'))
    """
    insert_seccion_sql = """
        INSERT OR REPLACE INTO layout_secciones (
            empresa_id, modulo, anio, capitulo,
            seccion_principal, seccion_secundaria, tipo, orden,
            creado_en, actualizado_en
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))
    """
    total = 0
    for capitulo, cuentas in grouped.items():
        principales = set()
        secundarias = set()
        for index, cuenta in enumerate(cuentas):
            principal = cuenta[KEY_PRINCIPAL_TARGET]
            secundaria = cuenta[KEY_SECUNDARIA_TARGET] or None
            conn.execute(
                insert_cuenta_sql,
                (
                    empresa_id,
                    modulo,
                    anio,
                    cuenta["CUENTA"],
                    cuenta["NOMBRE"],
                    capitulo,
                    principal,
                    secundaria,
                    index,
                ),
            )
            total += 1
            if principal and (capitulo, principal) not in principales:
                conn.execute(
                    insert_seccion_sql,
                    (
                        empresa_id,
                        modulo,
                        anio,
                        capitulo,
                        principal,
                        None,
                        "principal",
                        index,
                    ),
                )
                principales.add((capitulo, principal))
            if secundaria and (capitulo, principal, secundaria) not in secundarias:
                conn.execute(
                    insert_seccion_sql,
                    (
                        empresa_id,
                        modulo,
                        anio,
                        capitulo,
                        principal,
                        secundaria,
                        "secundaria",
                        index,
                    ),
                )
                secundarias.add((capitulo, principal, secundaria))
    return total


def insert_operaciones(conn, modulo, operaciones, empresa_id, anio):
    conn.execute(
        "DELETE FROM layout_operaciones WHERE empresa_id = ? AND modulo = ? AND anio = ?",
        (empresa_id, modulo, anio),
    )
    insert_sql = """
        INSERT INTO layout_operaciones (
            empresa_id, modulo, anio, capitulo, clase, seccion,
            operacion_tipo, operacion_label, signo, orden, actualizado_en
        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'))
    """
    for op in operaciones:
        conn.execute(
            insert_sql,
            (
                empresa_id,
                modulo,
                anio,
                op["capitulo"],
                op["clase"],
                op["seccion"],
                op["tipo"],
                op["label"],
                op["signo"],
                op["orden"],
            ),
        )
    return len(operaciones)


def import_module(conn, modulo, cuentas, operaciones, empresa_id, anio):
    total_cuentas = insert_cuentas(conn, modulo, cuentas, empresa_id, anio)
    total_operaciones = insert_operaciones(conn, modulo, operaciones, empresa_id, anio)
    print(
        f"[{modulo}] {total_cuentas} cuentas y {total_operaciones} operaciones cargadas para {anio}"
    )


def parse_args():
    parser = argparse.ArgumentParser(
        description="Importa los layouts de SUMMARY y RESUMEN hacia SQLite."
    )
    parser.add_argument(
        "--db",
        dest="db_path",
        default=str(DEFAULT_DB_PATH),
        help="Ruta al archivo panel.sqlite (default datos/panel.sqlite).",
    )
    parser.add_argument(
        "--cuentas",
        dest="cuentas_path",
        default=None,
        help="Archivo JSON con las cuentas de SUMMARY/RESUMEN.",
    )
    parser.add_argument(
        "--operaciones",
        dest="ops_path",
        default=None,
        help="Archivo JSON con la secci\u00f3n 'SUMA DE VARIAS SECCIONES'.",
    )
    parser.add_argument(
        "--empresa",
        dest="empresa_id",
        default=EMPRESA_ID,
        help="Empresa objetivo (default EMPRESA01).",
    )
    parser.add_argument(
        "--anio",
        dest="anio",
        default=str(TARGET_YEAR),
        help="A\u00f1o objetivo (default 2025).",
    )
    return parser.parse_args()


def ensure_rows(data, key, fallback_files):
    rows = data.get(key, [])
    if rows:
        return rows
    for candidate in fallback_files:
        path = (ROOT / candidate).resolve()
        if path.exists():
            return load_json(path).get(key, [])
    return []


def main():
    args = parse_args()
    db_path = Path(args.db_path).expanduser()
    anio = int(args.anio)
    empresa_id = args.empresa_id

    cuentas_path = resolve_path(args.cuentas_path, DEFAULT_CUENTAS)
    operaciones_path = resolve_path(args.ops_path, DEFAULT_OPERACIONES)

    cuentas_data = load_json(cuentas_path)
    operaciones_data = load_json(operaciones_path)["SUMA DE VARIAS SECCIONES"]

    summary_rows = ensure_rows(
        cuentas_data, "SUMMARY", ["info IMPORTANTE/CUENTAS SUMMARY.json"]
    )
    resumen_rows = ensure_rows(
        cuentas_data, "RESUMEN", ["info IMPORTANTE/CUENTAS SUMMARY y RESUMEN 2025.json"]
    )
    if not summary_rows:
        raise ValueError("No hay datos de SUMMARY en el archivo proporcionado.")
    if not resumen_rows:
        raise ValueError("No hay datos de RESUMEN en el archivo proporcionado.")

    summary_accounts = group_accounts(summary_rows)
    resumen_accounts = group_accounts(resumen_rows)
    summary_ops = build_operations(operaciones_data, "SUMMARY")
    resumen_ops = build_operations(operaciones_data, "RESUMEN")

    conn = sqlite3.connect(db_path)
    try:
        import_module(conn, "RESUMEN", resumen_accounts, resumen_ops, empresa_id, anio)
        import_module(conn, "SUMMARY", summary_accounts, summary_ops, empresa_id, anio)
        conn.commit()
    finally:
        conn.close()
    print(f"Base de datos actualizada: {db_path}")


if __name__ == "__main__":
    main()

File: scripts/import_summary_resumen_2025.js
const fs = require("fs");
const path = require("path");

const dbModule = require("../src/db/sqlite");
const { ahoraSql } = dbModule;

const ROOT = path.resolve(__dirname, "..");
const DEFAULT_CUENTAS = [
  "info IMPORTANTE/RESUMEN CUENTAS 2025.json",
  "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN 2025.json",
  "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN.json",
];
const DEFAULT_OPERACIONES = [
  "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN 2025.json",
  "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN.json",
];
const EMPRESA_ID = "EMPRESA01";
const TARGET_YEAR = 2025;

const KEY_PRINCIPAL_TARGET = "SECCI\u00e0N Principal";
const KEY_SECUNDARIA_TARGET = "SECCION Secundaria";
const PRINCIPAL_KEYS = [
  "SECCI\u00f3N Principal",
  "SECCI\u00e0N Principal",
  "SECCI\u00f3N",
  "SECCI\u00e0N",
  "SECCION Principal",
  "SECCION",
];
const SECUNDARIA_KEYS = [
  "SECCION Secundaria",
  "SECCI\u00f3N Secundaria",
  "SECCI\u00e0N Secundaria",
  "Seccion Secundaria",
  "SECCION",
  "SECCI\u00f3N",
  "SECCI\u00e0N",
];
const OPERATION_KEYS = [
  "sum-row",
  "sum-row-sumavarios",
  "sum-row-sumavarios-consolidado",
  "sum-row-operativo",
  "sum-row-operativo-consolidado",
  "result-row",
  "net-row",
  "net-row-adicional",
  "result-net-row",
];

const resolvePath = (value, fallbacks) => {
  if (value) {
    const candidate = path.resolve(value);
    if (fs.existsSync(candidate)) {
      return candidate;
    }
    throw new Error(`No existe el archivo ${candidate}`);
  }
  for (const fallback of fallbacks) {
    const candidate = path.resolve(ROOT, fallback);
    if (fs.existsSync(candidate)) {
      return candidate;
    }
  }
  throw new Error(
    "No se encontro archivo de entrada. Usa --cuentas o --operaciones para definirlo."
  );
};

const loadJson = (filePath) => {
  return JSON.parse(fs.readFileSync(filePath, "utf8"));
};

const cleanText = (value) => {
  if (value == null) return "";
  return String(value).trim();
};

const normalizePrincipal = (row) => {
  for (const key of PRINCIPAL_KEYS) {
    if (row[key]) {
      return cleanText(row[key]);
    }
  }
  return "";
};

const normalizeSecundaria = (row) => {
  for (const key of SECUNDARIA_KEYS) {
    if (row[key]) {
      return cleanText(row[key]);
    }
  }
  return "";
};

const groupAccounts = (rows) => {
  const grouped = {};
  for (const row of rows) {
    const capitulo = cleanText(row.CAPITULO);
    if (!capitulo) {
      continue;
    }
    const cuenta = cleanText(row.CUENTA);
    const nombre = cleanText(row.NOMBRE);
    if (!cuenta) {
      continue;
    }
    if (!grouped[capitulo]) {
      grouped[capitulo] = [];
    }
    grouped[capitulo].push({
      CUENTA: cuenta,
      NOMBRE: nombre,
      [KEY_PRINCIPAL_TARGET]: normalizePrincipal(row),
      [KEY_SECUNDARIA_TARGET]: normalizeSecundaria(row),
    });
  }
  return grouped;
};

const buildOperations = (rawOps, hojaExpected) => {
  const hojaUpper = hojaExpected.toUpperCase();
  const filtrados = rawOps.filter(
    (op) => cleanText(op.HOJA).toUpperCase() === hojaUpper
  );
  const operaciones = [];
  filtrados.forEach((op, idx) => {
    const capitulo = cleanText(op.CAPITULO);
    const clase = cleanText(op.Clase);
    const seccion = cleanText(op.SECCION);
    if (!capitulo || !clase || !seccion) {
      return;
    }
    const signo = clase.toLowerCase().includes("expense") ? -1 : 1;
    OPERATION_KEYS.forEach((tipo, tipoIdx) => {
      const etiqueta = cleanText(op[tipo]);
      if (!etiqueta) {
        return;
      }
      operaciones.push({
        capitulo,
        clase,
        seccion,
        tipo,
        label: etiqueta,
        signo,
        orden: idx * 100 + tipoIdx,
      });
    });
  });
  return operaciones;
};

const insertCuentas = async (db, modulo, grouped, empresaId, anio) => {
  await db.layout_cuentas.remove(
    { empresa_id: empresaId, modulo, anio },
    { multi: true }
  );
  await db.layout_secciones.remove(
    { empresa_id: empresaId, modulo, anio },
    { multi: true }
  );

  let total = 0;
  const now = ahoraSql();

  for (const [capitulo, cuentas] of Object.entries(grouped)) {
    const principales = new Set();
    const secundarias = new Set();

    for (const [index, cuenta] of cuentas.entries()) {
      const principal = cuenta[KEY_PRINCIPAL_TARGET];
      const secundaria = cuenta[KEY_SECUNDARIA_TARGET] || null;
      await db.layout_cuentas.upsertByKey({
        empresa_id: empresaId,
        modulo,
        anio,
        cuenta: cuenta.CUENTA,
        nombre: cuenta.NOMBRE,
        capitulo,
        seccion_principal: principal,
        seccion_secundaria: secundaria,
        orden: index,
        creado_en: now,
        actualizado_en: now,
      });
      total += 1;

      if (principal && !principales.has(`${capitulo}|${principal}`)) {
        principales.add(`${capitulo}|${principal}`);
        await db.layout_secciones.upsertByKey({
          empresa_id: empresaId,
          modulo,
          anio,
          capitulo,
          seccion_principal: principal,
          seccion_secundaria: null,
          tipo: "principal",
          orden: index,
          creado_en: now,
          actualizado_en: now,
        });
      }

      if (secundaria) {
        const key = `${capitulo}|${principal}|${secundaria}`;
        if (!secundarias.has(key)) {
          secundarias.add(key);
          await db.layout_secciones.upsertByKey({
            empresa_id: empresaId,
            modulo,
            anio,
            capitulo,
            seccion_principal: principal,
            seccion_secundaria: secundaria,
            tipo: "secundaria",
            orden: index,
            creado_en: now,
            actualizado_en: now,
          });
        }
      }
    }
  }

  return total;
};

const insertOperaciones = async (db, modulo, operaciones, empresaId, anio) => {
  await db.layout_operaciones.remove(
    { empresa_id: empresaId, modulo, anio },
    { multi: true }
  );
  const now = ahoraSql();

  for (const op of operaciones) {
    await db.layout_operaciones.upsertByKey({
      empresa_id: empresaId,
      modulo,
      anio,
      capitulo: op.capitulo,
      clase: op.clase,
      seccion: op.seccion,
      operacion_tipo: op.tipo,
      operacion_label: op.label,
      operacion_etiqueta: op.clase,
      signo: op.signo,
      orden: op.orden,
      creado_en: now,
      actualizado_en: now,
    });
  }

  return operaciones.length;
};

const importModule = async (db, modulo, cuentas, operaciones, empresaId, anio) => {
  const totalCuentas = await insertCuentas(db, modulo, cuentas, empresaId, anio);
  const totalOperaciones = await insertOperaciones(
    db,
    modulo,
    operaciones,
    empresaId,
    anio
  );
  console.log(
    `[${modulo}] ${totalCuentas} cuentas y ${totalOperaciones} operaciones cargadas para ${anio}`
  );
};

const parseArgs = () => {
  const args = process.argv.slice(2);
  const getValue = (flag) => {
    const idx = args.indexOf(`--${flag}`);
    if (idx >= 0 && idx < args.length - 1) {
      return args[idx + 1];
    }
    return null;
  };
  return {
    cuentasPath: getValue("cuentas"),
    opsPath: getValue("operaciones"),
    empresaId: getValue("empresa") || EMPRESA_ID,
    anio: Number(getValue("anio") || TARGET_YEAR),
  };
};

const ensureRows = (data, key, fallbackFiles) => {
  const rows = Array.isArray(data?.[key]) ? data[key] : [];
  if (rows.length) {
    return rows;
  }
  for (const candidate of fallbackFiles) {
    const filePath = path.resolve(ROOT, candidate);
    if (fs.existsSync(filePath)) {
      const contenido = loadJson(filePath);
      const alterno = Array.isArray(contenido?.[key]) ? contenido[key] : [];
      if (alterno.length) {
        return alterno;
      }
    }
  }
  return [];
};

const main = async () => {
  const args = parseArgs();
  const anio = Number.isFinite(args.anio) ? args.anio : TARGET_YEAR;

  const cuentasPath = resolvePath(args.cuentasPath, DEFAULT_CUENTAS);
  const operacionesPath = resolvePath(args.opsPath, DEFAULT_OPERACIONES);

  const cuentasData = loadJson(cuentasPath);
  const operacionesData = loadJson(operacionesPath);
  const operacionesRaw = operacionesData["SUMA DE VARIAS SECCIONES"] || [];

  const summaryRows = ensureRows(cuentasData, "SUMMARY", [
    "info IMPORTANTE/CUENTAS SUMMARY.json",
  ]);
  const resumenRows = ensureRows(cuentasData, "RESUMEN", [
    "info IMPORTANTE/CUENTAS SUMMARY y RESUMEN 2025.json",
  ]);

  if (!summaryRows.length) {
    throw new Error("No hay datos de SUMMARY en el archivo proporcionado.");
  }
  if (!resumenRows.length) {
    throw new Error("No hay datos de RESUMEN en el archivo proporcionado.");
  }

  const summaryAccounts = groupAccounts(summaryRows);
  const resumenAccounts = groupAccounts(resumenRows);
  const summaryOps = buildOperations(operacionesRaw, "SUMMARY");
  const resumenOps = buildOperations(operacionesRaw, "RESUMEN");

  await dbModule.inicializarBaseDatos();
  const layoutService = require("../src/services/layoutService");
  const empresaCanonica = layoutService.obtenerEmpresaCanonica(args.empresaId);
  const { db } = dbModule;

  await importModule(db, "RESUMEN", resumenAccounts, resumenOps, empresaCanonica, anio);
  await importModule(db, "SUMMARY", summaryAccounts, summaryOps, empresaCanonica, anio);

  console.log(`Base de datos actualizada para ${empresaCanonica} (${anio}).`);
};

if (require.main === module) {
  main().catch((error) => {
    console.error("Error importando layouts:", error.message || error);
    process.exit(1);
  });
}
